### PR TITLE
Increase precision to high for cpp/static-buffer-overflow

### PR DESCRIPTION
**Pull request by andersfugmann**
**As this changes the precision on static-buffer-overflow query, please do not merge before GHES 3.3 has been branched.**

This PR includes:
* Further relax predicate `memberMaybeVarsized` to include all members of size 0 or 1
* Remove warnings for static buffer overflows in when fields are indexed as an argument to `offsetof`
* Do not warn on static buffer overflows if upper bound analysis used widening. 
* Weaken wording on overflow static alert text (`is accessed` -> `may be accessed`) 

Statistics from running OverflowStatic on all LGTM projects:
|  | |
| ----------- | ----------- |
| Total Alerts | 206 | 
| Projects with alerts | 91 | 
| False positives | 16 | 
| Unknown* | 6 | 

[*] Unknown alerts denotes alerts, where I cannot judge if the alerts is correct. These usually includes nested macro expansions. 

The false positives can be grouped into:
* Copying multiple fields (7 FP's)
* Limitations in the range analysis (1 FP)
  * 1 reasonably trivial 
  * 3 are non trivial, and would require a deeper invariant analysis. 
* Dead code (2 FP's)
* Wrong assumption of `page_size` (evaluated to 0) (1 FP)
* Variable sized member of size 4 (`char data[sizeof(int)]`) (2 FP)

Detailed results and the actual analysis can be found here: 
https://gist.github.com/andersfugmann/5091e7c63cd4f6ae94d3b86e426efd77

### CODE REVIEW

File: cpp/change-notes/2021-09-27-overflow-static.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+lgtm,codescanning
+* Increase presition to high for "Static buffer overflow" query
+  (cpp/static-buffer-overflow).
```
Comment (jbj) [inline cpp/change-notes/2021-09-27-overflow-static.md:3]:
```suggestion
* Increase precision to high for the "Static buffer overflow" query
  (`cpp/static-buffer-overflow`). This means the query is run and displayed by default on Code Scanning and LGTM.
```

File: cpp/ql/lib/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll [hunk 2]
```diff
@@ -1593,6 +1593,14 @@ private module SimpleRangeAnalysisCached {
     result = min([max(getTruncatedUpperBounds(expr)), getGuardedUpperBound(expr)])
   }
 
+  /** Holds if `expr` may have been widened */
```
Comment (jbj) [inline cpp/ql/lib/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll:1596]:
```suggestion
  /** Holds if the upper bound of `expr` may have been widened. This means the the upper bound is in practice likely to be overly wide. */
```

File: cpp/ql/lib/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll [hunk 3]
```diff
@@ -1593,6 +1593,14 @@ private module SimpleRangeAnalysisCached {
...
+  cached
+  predicate upperBoundMayBeWidened(Expr e) {
+    isRecursiveExpr(e) and
+    // Corresponds to taking max on the RHS
+    not getGuardedUpperBound(e) < getTruncatedUpperBounds(e)
```
Comment (jbj) [inline cpp/ql/lib/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll:1601]:
```suggestion
    // Widening is not a problem if the post-analysis in `getGuardedUpperBound` has overridden the widening.
    // Note that the RHS of `<` may be multi-valued.
    not getGuardedUpperBound(e) < getTruncatedUpperBounds(e)
```

**Review by jbj — COMMENTED** [review]
The code changes LGTM, but I've suggested some expansion of the docs.

This PR removes many good results in the tests, but I think those results were on contrived code.

**Comment by andersfugmann** [comment]
Test fails on `/home/runner/work/semmle-code/semmle-code/semmlecode-cpp-tests/DO_NOT_DISTRIBUTE/security-tests/CWE-120/CWE/examples/OverrunWrite.qlref`. Any suggestions on how I can resolve this?

**Comment by MathiasVP** [comment]
> Test fails on `/home/runner/work/semmle-code/semmle-code/semmlecode-cpp-tests/DO_NOT_DISTRIBUTE/security-tests/CWE-120/CWE/examples/OverrunWrite.qlref`. Any suggestions on how I can resolve this?

These needs to be resolved in the internal repo. You need to create a PR in the internal repo with the changes and bump the QL submodule in the PR to point to the HEAD of this PR.

**Comment by jbj** [comment]
I'm around to help in the office.

**Review by MathiasVP — APPROVED** [review]
LGTM!

**Pull request merged**